src/Menu/menu/sizeModifier.ts

```diff
@@ -0,0 +1,292 @@
+
+function getPlacementAndMaxSize({
+  state,
+  options,
+}: ModifierArguments<Options>): {
+  placement: Placement;
+  maxSize: { height: number };
+} {
+  const {
+    fallbackPlacements: specifiedFallbackPlacements,
+    padding = getPaddingFromState(state),
+    boundary = "clippingParents",
+    rootBoundary = "viewport",
+    altBoundary,
+    flipVariations = true,
+    allowedAutoPlacements,
+  } = options;
+
+  /**
+   * Preferred placement
+   *
+   * Taken from options
+   */
+  const preferredPlacement = state.options.placement;
+
+  /**
+   * Base placement from `preferredPlacement`
+   */
+  const basePlacement = getBasePlacement(preferredPlacement);
+
+  /**
+   * Represents if the `preferredPlacement` is a `BasePlacement` (meaning it has
+   * no `-begin` or `-end`)
+   */
+  const isBasePlacement = basePlacement === preferredPlacement;
+
+  /**
+   * List of fallback placements
+   *
+   * Either passed in with `specifiedFallbackPlacements`, or calculated
+   * depending on
+   * [`flipVariations`](https://popper.js.org/docs/v2/modifiers/flip/#flipvariations)
+   */
+  const fallbackPlacements =
+    specifiedFallbackPlacements ||
+    (isBasePlacement || !flipVariations
+      ? [getOppositePlacement(preferredPlacement)]
+      : getExpandedFallbackPlacements(preferredPlacement));
+
+  /**
+   * Aggregate of all placements, including preferred and fallbacks.
+   */
+  const placements = [preferredPlacement, ...fallbackPlacements].reduce<
+    Placement[]
+  >((acc, placement) => {
+    return acc.concat(
+      getBasePlacement(placement) === "auto"
+        ? computeAutoPlacement(state, {
+            placement,
+            boundary,
+            rootBoundary,
+            padding,
+            flipVariations,
+            allowedAutoPlacements,
+          })
+        : placement
+    );
+  }, []);
+
+  const referenceRect = state.rects.reference;
+  const popperRect = state.rects.popper;
+
+  /**
+   * Array calculated from `placements` with the calculated values of each
+   * `placement`. This will be used to determine if we're capable of flipping
+   * the element to display it ir of we have to set the `max-height` too.
+   */
+  const overflows = placements.map<OverflowDict>((placement) => {
+    const basePlacement = getBasePlacement(placement);
+    const isStartVariation = getVariation(placement) === start;
+    const isVertical = basePlacement === "top" || basePlacement === "bottom";
+    const len = isVertical ? "width" : "height";
+
+    const overflow = detectOverflow(state, {
+      placement,
+      boundary,
+      rootBoundary,
+      altBoundary,
+      padding,
+    });
+
+    let mainVariationSide: BasePlacement = isVertical
+      ? isStartVariation
+        ? "right"
+        : "left"
+      : isStartVariation
+      ? "bottom"
+      : "top";
+
+    if (referenceRect[len] > popperRect[len]) {
+      mainVariationSide = (getOppositePlacement(
+        mainVariationSide
+      ) as any) as BasePlacement;
```

Comment:
You shouldnt need the as any here, afaict `Placement` should be assignable to `BasePlacement`?

Reply:
`Placement` is not assignable to `BasePlacement`, `BasePlacement` is assignable to `Placement`. In other words, `BasePlacement` is a subset of `Placement`. There's no way for `getOppositePlacement` to know its input is a `BasePlacement` and, therefore, its output would be the same. It would make sense to add a generic to `getOppositePlacement`, but I opted instead fo create `getOppositeBasePlacement` so there's no confusion or typing issues.

Reply:
Sorry I meant that the other way around. I agree you still need the `as BasePlacement` here, but not the `as any`